bioio_ome_zarr/writers/ome_zarr_writer.py

```diff
@@ -0,0 +1,489 @@
+                    )
+                self.chunk_shapes_per_level = [
+                    tuple(int(x) for x in single) for _ in range(self.num_levels)
+                ]
+            else:
+                per_level = cast(ChunkPerLevel, chunk_shape)
+                if len(per_level) != self.num_levels:
+                    raise ValueError(
+                        f"chunk_shape must match levels: ({self.num_levels}); "
+                        f"got {len(per_level)}"
+                    )
+                shapes: List[Tuple[int, ...]] = []
+                for idx, ch in enumerate(per_level):
+                    if len(ch) != self.ndim:
+                        raise ValueError(
+                            f"chunk_shape[{idx}] length {len(ch)} != ndim {self.ndim}"
+                        )
+                    shapes.append(tuple(int(x) for x in ch))
+                self.chunk_shapes_per_level = shapes
+        else:
+            suggested = chunk_size_from_memory_target(
+                self.level_shapes[0], self.dtype, 16 << 20
+            )
+            self.chunk_shapes_per_level = [suggested for _ in range(self.num_levels)]
+
+        # 5) formatting and compression
+        self.zarr_format = zarr_format
+        self.shard_factor = shard_factor
+        self.compressor = compressor
+
+        # 6) Metadata fields
+        self.image_name = image_name or "Image"
+        self.channels = channels
+        self.rdefs = rdefs
+        self.creator_info = creator_info
+        self.root_transform = root_transform
+
+        # 8) Handles & state
+        self.root: Optional[zarr.Group] = None
+        self.datasets: List[zarr.Array] = []
+        self._initialized: bool = False
+        self._metadata_written: bool = False
+
+    # -----------------
+    # Public interface
+    # -----------------
+
+    def write_full_volume(
+        self,
+        input_data: Union[np.ndarray, da.Array],
+    ) -> None:
+        """
+        Write full-resolution data into all pyramid levels.
+
+        Parameters
+        ----------
+        input_data : Union[np.ndarray, dask.array.Array]
+            Array matching level-0 shape. If NumPy, it will be wrapped into a
+            Dask array with level-0 chunking.
+        """
+        if not self._initialized:
+            self._initialize()
+
+        base = (
+            input_data
+            if isinstance(input_data, da.Array)
+            else da.from_array(input_data, chunks=self.datasets[0].chunks)
+        )
+
+        # Store each level (downsampled with nearest-neighbor for parity)
+        for lvl, shape in enumerate(self.level_shapes):
+            src = base if lvl == 0 else resize(base, shape, order=0)
+            if self.zarr_format == 2:
+                da.to_zarr(src, self.datasets[lvl])
+            else:
+                da.store(src, self.datasets[lvl], lock=True)
+
+    def write_timepoints(
+        self,
+        source: Union[Reader, np.ndarray, da.Array],
```

Comment:
Wondering if `Reader` should just be removed? Its a nice to have but is it necessary & adding confusion?

Reply:
yeah im for it.